edx_notifications/data.py

```diff
@@ -0,0 +1,169 @@
+"""
+All pure data objects that the various Data Providers will product. This is help avoid
+generic dictionaries from being passed around, plus this will help avoid any
+implicit database-specific bindings that come with any uses of ORMs.
+"""
+
+from django.core.exceptions import ValidationError
+
+from edx_notifications import const
+
+from edx_notifications.base_data import (
+    StringField,
+    DictField,
+    DateTimeField,
+    IntegerField,
+    EnumField,
+    RelatedObjectField,
+    BaseDataObject
+)
+
+
+class NotificationChannel(BaseDataObject):
+    """
+    Specifies a channel through which a notification is delivered
+    """
+
+    # the internal name of the channel
+    name = StringField()
+
+    # the human readible name of the channel
+    display_name = StringField()
+
+    # the human readible description of the channel
+    description_name = StringField()
+
+
+class NotificationType(BaseDataObject):
+    """
+    The Data Object representing the NotificationType
+    """
+
+    # the name (including namespace) of the notification, e.g. open-edx.lms.forums.reply-to-post
+    name = StringField()
+
+    # the human readible string of the name of the notification
+    display_name = StringField()
+
+    # the human readible string that describes the notification
+    display_description = StringField()
+
+    # default delivery channel for this type
+    # None = no default
+    default_channel = RelatedObjectField(NotificationChannel)
+
+    # renderer class - as a string - that will handle the rendering of
+    # this type
+    renderer = StringField()
```

Comment:
Well, different NotificationTypes will have different payload schema's, for example, a notification regarding a forum posting will need to present a message regarding the Topic and who created it, etc.. That is why I'm binding a Renderer to a NotificationType. Multiple NotificationTypes can point to the same renderer, for re-use. 

In order to account for different channels needing different renderings, there's a "format" property that is passed into the renderer. You can see an example in edx_notifications/renderers/base.py which is a simple Subject/Body renderer which - for now - only allows for format="underscore" (for client-side rendering).

Let me see if I can fill in some detail and provide more sample code for you.